🌊 LLM-powered parsing suggestions

flash1293 opened the pull request:
Depends on https://github.com/elastic/kibana/pull/209985

Add suggestions for grok processing:

<img width="594" alt="Screenshot 2025-02-05 at 10 31 27" src="https://github.com/user-attachments/assets/4b717681-aa7d-4952-a4e0-9013d9b8aaf8" />

The logic for generating suggestions works like this:
* Take the current sample
* Split it into patterns based on a simple regex-based grouping replacing runs of numbers with a placeholder, runs of regular numbers with a placeholder, etc.
* For the top 5 found groups, pass a couple messages to the LLM in parallel to come up with a grok pattern
* Check the grok patterns whether they actually match something and don't break
* Report the patterns that have a positive match rate

For the `Generate patterns` button to show in the UI, make sure a connector is configured and the license level is above basic (trial license is easiest to test with).

I did some light refactoring on the processing routes, moving the simulation bits into a separate file - no changes in this area though.

<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/273439)
* Commit: 2aa13313613af78e6f9afb2d2227169fedd5262b

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/273439#0194d58e-0f76-4130-a58b-5cf3ee30f72f)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273439","number":273439,"commit":"2aa13313613af78e6f9afb2d2227169fedd5262b"}],"number":273439}
buildkite-pr-comment-->

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/273515)
* Commit: bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/273515#0194d60d-7491-4333-b267-d6eec7be0a55)

### History
* :broken_heart: [Build #273439](https://buildkite.com/elastic/kibana-pull-request/builds/273439) failed 2aa13313613af78e6f9afb2d2227169fedd5262b

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273515","number":273515,"commit":"bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273439","number":273439,"commit":"2aa13313613af78e6f9afb2d2227169fedd5262b"}],"number":273515}
buildkite-pr-comment-->

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/273545)
* Commit: c70cff03200ca3fc3a50f13d4c31b3f60e509a64
* Kibana Serverless Image: `docker.elastic.co/kibana-ci/kibana-serverless:pr-208777-c70cff03200c`

### History
* :broken_heart: [Build #273515](https://buildkite.com/elastic/kibana-pull-request/builds/273515) failed bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152
* :broken_heart: [Build #273439](https://buildkite.com/elastic/kibana-pull-request/builds/273439) failed 2aa13313613af78e6f9afb2d2227169fedd5262b

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273545","number":273545,"commit":"c70cff03200ca3fc3a50f13d4c31b3f60e509a64"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273515","number":273515,"commit":"bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273439","number":273439,"commit":"2aa13313613af78e6f9afb2d2227169fedd5262b"}],"number":273545}
buildkite-pr-comment-->

kuisathaverat commented:
/oblt-deploy

tonyghiani commented on src/platform/packages/shared/kbn-zod-helpers/src/deep_strict.ts:
```diff
@@ -311,11 +311,11 @@ export function DeepStrict<TSchema extends z.Schema>(schema: TSchema) {
-        keys: excessKeys,
-        message: `Excess keys are not allowed`,
-      });
+      // context.addIssue({
+      //   code: ZodIssueCode.unrecognized_keys,
+      //   keys: excessKeys,
+      //   message: `Excess keys are not allowed`,
+      // });
```
The changes in [[Streams 🌊] Improve typing for samples and simulation docs](https://github.com/elastic/kibana/pull/209991#top) should fix this part to which can be uncommented once it's merged.

flash1293 commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274220)
* Commit: fa4b7608d1c2f8ac39b4f02ac6c375c9a1c25f6b

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/274220#0194dfc7-dd3c-4966-9089-412b570bf838)

### History
* :broken_heart: [Build #273545](https://buildkite.com/elastic/kibana-pull-request/builds/273545) failed c70cff03200ca3fc3a50f13d4c31b3f60e509a64
* :broken_heart: [Build #273515](https://buildkite.com/elastic/kibana-pull-request/builds/273515) failed bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152
* :broken_heart: [Build #273439](https://buildkite.com/elastic/kibana-pull-request/builds/273439) failed 2aa13313613af78e6f9afb2d2227169fedd5262b

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274220","number":274220,"commit":"fa4b7608d1c2f8ac39b4f02ac6c375c9a1c25f6b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273545","number":273545,"commit":"c70cff03200ca3fc3a50f13d4c31b3f60e509a64"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273515","number":273515,"commit":"bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273439","number":273439,"commit":"2aa13313613af78e6f9afb2d2227169fedd5262b"}],"number":274220}
buildkite-pr-comment-->

flash1293 commented:
/ci

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274303)
* Commit: a81de65acb3659708cdab24bf354a6ca65c91516

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/274303#0194e09e-deb6-4be9-81b8-be18fc502f92)

### History
* :broken_heart: [Build #274220](https://buildkite.com/elastic/kibana-pull-request/builds/274220) failed fa4b7608d1c2f8ac39b4f02ac6c375c9a1c25f6b
* :broken_heart: [Build #273545](https://buildkite.com/elastic/kibana-pull-request/builds/273545) failed c70cff03200ca3fc3a50f13d4c31b3f60e509a64
* :broken_heart: [Build #273515](https://buildkite.com/elastic/kibana-pull-request/builds/273515) failed bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152
* :broken_heart: [Build #273439](https://buildkite.com/elastic/kibana-pull-request/builds/273439) failed 2aa13313613af78e6f9afb2d2227169fedd5262b

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274303","number":274303,"commit":"a81de65acb3659708cdab24bf354a6ca65c91516"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274220","number":274220,"commit":"fa4b7608d1c2f8ac39b4f02ac6c375c9a1c25f6b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273545","number":273545,"commit":"c70cff03200ca3fc3a50f13d4c31b3f60e509a64"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273515","number":273515,"commit":"bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273439","number":273439,"commit":"2aa13313613af78e6f9afb2d2227169fedd5262b"}],"number":274303}
buildkite-pr-comment-->

flash1293 commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274356)
* Commit: 47dd7ed5d4dedcd58147dfd030ab78af8f4dad1e

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/274356#0194e0fb-b588-490e-b905-74d850cc8af6)

### History
* :broken_heart: [Build #274303](https://buildkite.com/elastic/kibana-pull-request/builds/274303) failed a81de65acb3659708cdab24bf354a6ca65c91516
* :broken_heart: [Build #274220](https://buildkite.com/elastic/kibana-pull-request/builds/274220) failed fa4b7608d1c2f8ac39b4f02ac6c375c9a1c25f6b
* :broken_heart: [Build #273545](https://buildkite.com/elastic/kibana-pull-request/builds/273545) failed c70cff03200ca3fc3a50f13d4c31b3f60e509a64
* :broken_heart: [Build #273515](https://buildkite.com/elastic/kibana-pull-request/builds/273515) failed bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152
* :broken_heart: [Build #273439](https://buildkite.com/elastic/kibana-pull-request/builds/273439) failed 2aa13313613af78e6f9afb2d2227169fedd5262b

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274356","number":274356,"commit":"47dd7ed5d4dedcd58147dfd030ab78af8f4dad1e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274303","number":274303,"commit":"a81de65acb3659708cdab24bf354a6ca65c91516"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274220","number":274220,"commit":"fa4b7608d1c2f8ac39b4f02ac6c375c9a1c25f6b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273545","number":273545,"commit":"c70cff03200ca3fc3a50f13d4c31b3f60e509a64"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273515","number":273515,"commit":"bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273439","number":273439,"commit":"2aa13313613af78e6f9afb2d2227169fedd5262b"}],"number":274356}
buildkite-pr-comment-->

flash1293 commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274368)
* Commit: 0695680f05f93c0d217c5672077c89c21e0153d2
* Kibana Serverless Image: `docker.elastic.co/kibana-ci/kibana-serverless:pr-208777-0695680f05f9`

### History
* :broken_heart: [Build #274356](https://buildkite.com/elastic/kibana-pull-request/builds/274356) failed 47dd7ed5d4dedcd58147dfd030ab78af8f4dad1e
* :broken_heart: [Build #274303](https://buildkite.com/elastic/kibana-pull-request/builds/274303) failed a81de65acb3659708cdab24bf354a6ca65c91516
* :broken_heart: [Build #274220](https://buildkite.com/elastic/kibana-pull-request/builds/274220) failed fa4b7608d1c2f8ac39b4f02ac6c375c9a1c25f6b
* :broken_heart: [Build #273545](https://buildkite.com/elastic/kibana-pull-request/builds/273545) failed c70cff03200ca3fc3a50f13d4c31b3f60e509a64
* :broken_heart: [Build #273515](https://buildkite.com/elastic/kibana-pull-request/builds/273515) failed bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274368","number":274368,"commit":"0695680f05f93c0d217c5672077c89c21e0153d2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274356","number":274356,"commit":"47dd7ed5d4dedcd58147dfd030ab78af8f4dad1e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274303","number":274303,"commit":"a81de65acb3659708cdab24bf354a6ca65c91516"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274220","number":274220,"commit":"fa4b7608d1c2f8ac39b4f02ac6c375c9a1c25f6b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273545","number":273545,"commit":"c70cff03200ca3fc3a50f13d4c31b3f60e509a64"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273515","number":273515,"commit":"bba2eb930358ffdbdd27bd9b9c7b7d15ad03a152"}],"number":274368}
buildkite-pr-comment-->

flash1293 commented on x-pack/solutions/observability/packages/kbn-streams-schema/src/helpers/condition_to_query_dsl.ts:
```diff
@@ -62,6 +63,9 @@ export function conditionToQueryDsl(condition: Condition): any {
       },
     };
   }
+  if (isAlwaysCondition(condition)) {
```
This is a general fix I stumbled over - right now `always: {}` translates to `match_none: {}`, which is very confusing.

dgieselaar commented on x-pack/solutions/observability/plugins/streams/kibana.jsonc:
```diff
@@ -17,7 +17,8 @@
       "taskManager",
-      "alerting"
+      "alerting",
+      "inference",
```
not really related to this PR, but that looks like a bunch of dependencies that we can remove? taskManager, encryptedSavedObject, maybe even alerting although I guess we have some references to that in the AssetService

dgieselaar commented on x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/route.ts:
```diff
@@ -35,7 +32,7 @@ const paramsSchema = z.object({
 });
 
-type ProcessingSimulateParams = z.infer<typeof paramsSchema>;
+export type ProcessingSimulateParams = z.infer<typeof paramsSchema>;
```
not a biggie, but would like us to get into the habit of using `= z.ZodSchema<ExplicitType>`

dgieselaar replied:
also should we move this into the schema package then?

flash1293 replied:
+1 for z.Schema, but I think we don't need to move yet because it's only used internally in the server, not from the client.

dgieselaar commented on x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/simulation_handler.ts:
```diff
@@ -0,0 +1,205 @@
+    _source: doc,
+  }));
+
+  const simulationBody: any = {
```
is `any` here needed?

dgieselaar commented on x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/simulation_handler.ts:
```diff
@@ -0,0 +1,205 @@
+
+export const prepareSimulationBody = (params: ProcessingSimulateParams) => {
+  const { path, body } = params;
+  // eslint-disable-next-line @typescript-eslint/naming-convention
```
I'm fine with disabling this linting rule entirely for this plugin, but if we don't, maybe we can just use `detected_fields: detectedFields`?

dgieselaar commented on x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/simulation_handler.ts:
```diff
@@ -0,0 +1,205 @@
+export const executeSimulation = async (
+  scopedClusterClient: IScopedClusterClient,
+  simulationBody: ReturnType<typeof prepareSimulationBody>
+): Promise<any> => {
```
would prefer an explicit type here even if temporary

dgieselaar commented on x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/simulation_handler.ts:
```diff
@@ -0,0 +1,205 @@
+      body: simulationBody,
+    });
+  } catch (error) {
+    throw new SimulationFailedError(error);
```
do we need to catch/rethrow error here, instead of just propagating? I'm not sure if it applies here, but I think it's a good guiding principle to delegate to ES (including error handling) as much as we can, and then handle in the UI

dgieselaar commented on x-pack/solutions/observability/plugins/streams_app/public/components/stream_detail_enrichment/processors/index.tsx:
```diff
@@ -44,11 +44,13 @@ import { UseProcessingSimulatorReturn } from '../hooks/use_processing_simulator'
 export interface AddProcessorPanelProps extends ProcessorPanelProps {
   isInitiallyOpen?: boolean;
   onAddProcessor: UseDefinitionReturn['addProcessor'];
+  samples: RecursiveRecord[];
```
thoughts about making this an explicit type like `SampleDocument` so we don't start using this type in any place where it's easy? it's essentially `any`

tonyghiani replied:
> it's essentially any

What do you mean by this? The type is for typing nested structures with some limitations (we don't expect arrays of objects, hence that's not allowed) and we could get nested/flattened objects from the sampling. 
Do you have anything specific in mind for the `SampleDocument` type?

dgieselaar replied:
not sure if I follow, where would you keep it then?

tonyghiani replied:
What I mean is that we use the `RecursiveRecord` type might be used in other places (such as the simulation API) to type docs `_source` without them necessarily represent sample documents. I guess it would be confusing to use the SampleDocument type for something that it's not used as a sample.

flash1293 replied:
I will move the `RecursiveRecord` type and schema to `kbn-schema`, then rename all the usages to `SampleDocument`. Once we need `RecursiveRecord` somewhere else, we can easily split it out as a standalone thing, but let's not overcomplicate from the start.

dgieselaar commented on x-pack/solutions/observability/plugins/streams_app/public/components/stream_detail_enrichment/processors/grok/grok_ai_suggestions.tsx:
```diff
@@ -0,0 +1,322 @@
+          {i18n.translate(
+            'xpack.streams.streamDetailView.managementTab.enrichment.processorFlyout.refreshSuggestions',
+            {
+              defaultMessage: '✨ Generate patterns',
```
you can also consider using `sparkles` as the icon for the button, that might work better with different themes / dark mode/accessibility etc

dgieselaar commented on x-pack/solutions/observability/plugins/streams_app/public/components/stream_detail_enrichment/processors/grok/grok_ai_suggestions.tsx:
```diff
@@ -0,0 +1,322 @@
+
+  const abortController = useAbortController();
+
+  const refreshSuggestions = useCallback(() => {
```
is it not possible to use `useStreamsAppFetch()` here?

flash1293 replied:
`useStreamsAppFetch` triggers right away, but this should be on-demand only, which I didn't find a nice way to achieve on the consumer side. I pondered adding this functionality to `useAbortableAsync` (some kind of `fetchOnMount` option), but it felt like overloading it too much. Maybe I missed an easy way to achieve this, happy to improve.

dgieselaar replied:
I guess it would be something like `const [ refreshId, setRefreshId ] = useState(0);`

and then:

```ts

setRefreshId(prev => prev+1);

...

if (refreshId === 0) {
	return;
}
```

Not very pretty but beats re-implementing all the state management I guess? Is there no other state to use to determine when to fetch? (e.g. if the form values have changed from the original form values)

flash1293 replied:
Hmm, but I never want the suggestions to be triggered without the user clicking the button. It seems like the wrong tool for the job - not sure how to effectively abstract it out, maybe we should create a `useStreamsFetchCallback` which returns `{ value, isLoading, error, fetch }` or so. Wdyt?

dgieselaar replied:
yeah, that's probably better, but fine to postpone that as well. that probably could be captured in `useAbortableAsyncCallback`

tonyghiani replied:
You could have probably used the `useAsyncFn` hook here, as it abstract all the part for error handling and loading state and gives you the explicit trigger to send the request. But I wouldn't mind if this doesn't update now, with the state management refactor this will be handled at the processor level to prevent impossible side effects.

tonyghiani commented on x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/simulation_handler.ts:
@dgieselaar all the changes in this file were only moved from the `route.ts` file and don't really affect the scope of this work, I guess @flash1293 moved them as the route file was getting quite a lot of logic defined.
Regrding all the comments you have on this file, they are all good points, but I wouldn't spend time on updating them as everything about this endpoint id changing in the WIP [[Streams 🌊] Enrichment simulation behaviour improvements](https://github.com/elastic/kibana/pull/209985#top)

dgieselaar commented on x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/suggestions_handler.ts:
```diff
@@ -0,0 +1,235 @@
+) => {
+  const { field, samples } = body;
+  // Step 1: EVAL pattern
+  const evalPattern = (sample: string) => {
+    return sample
+      .replace(/[ \t\n]+/g, ' ')
+      .replace(/[A-Za-z]+/g, 'a')
+      .replace(/[0-9]+/g, '0')
+      .replace(/(a a)+/g, 'a')
+      .replace(/(a0)+/g, 'f')
+      .replace(/(f:)+/g, 'f:')
+      .replace(/0(.0)+/g, 'p');
+  };
```
maybe some comments/examples here?

flash1293 replied:
Right, I meant to add a unit test for this whole handler as it's doing a bunch of things and forgot - will add

tonyghiani commented on x-pack/solutions/observability/plugins/streams_app/public/components/stream_detail_enrichment/processors/grok/grok_patterns_editor.tsx:
```diff
@@ -58,42 +71,56 @@ export const GrokPatternsEditor = () => {
+          </SortableList>
+        </EuiPanel>
+      </EuiFormRow>
+      {isAiEnabled && refreshSimulation && definition && samples ? (
+        <GrokAiSuggestions
+          definition={definition}
+          refreshSimulation={refreshSimulation}
+          samples={samples}
+          extraButtons={addButton}
+        />
+      ) : (
+        addButton
       )}
```
I believe the Ai suggestions here shouldn't be coupled to this component specifically, but added at the GrokProcessorForm component level. This is a component that should purely handle the form patterns value, and doesn't need to know about the samples nor the definition.

tonyghiani replied:
Here is where we start getting more complex state to manage also suggestions related to the simulation itself the form changes. Introducing a context would work, and the plan was to also manage it with [[Streams] Enrichment - Migrate to solid state management system for stream app #102](https://github.com/elastic/streams-program/issues/102) re-work for maintainability.

For this part now, context looks good, although I'm not 100% convinced about the structure this form is taking.

flash1293 replied:
I'll go with context for now, I agree that we anyway want to rework the state management here in a follow-up.

tonyghiani commented on x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/suggestions_handler.ts:
```diff
@@ -0,0 +1,235 @@
+import { InferenceClient } from '@kbn/inference-plugin/server';
+import { RecursiveRecord } from '@kbn/streams-schema';
+import { ProcessingSuggestionParams } from './route';
+import {
+  assertSimulationResult,
+  executeSimulation,
+  prepareSimulationBody,
+  prepareSimulationDiffs,
+  prepareSimulationResponse,
+} from './simulation_handler';
```
@flash1293 all of this will be gone and won't adapt to the updated simulation API. Relying on them would probably break this API, I suggest to copy them over and later on adapt this API to use the new simulation logic where necessary.

flash1293 replied:
Sounds good, let's do this - I can either adjust here or you can on your side, depending on what merges first

tonyghiani replied:
Has this been applied? These utils should be copied here and not imported from the simulation_handler (since they won't exist anymore with https://github.com/elastic/kibana/pull/209985)

flash1293 commented:
Thanks for the comments, will fix! @dgieselaar I took the freedom to resolve your comments in the simulation handler as @tonyghiani will take care of this separately.

dgieselaar commented on x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/suggestions_handler.ts:
```diff
@@ -0,0 +1,235 @@
+      .replace(/0(.0)+/g, 'p');
+  };
+
+  const inputPatterns = samples.map((sample) => ({
```
is this (until L91) equivalent to:

```ts
  const samplesWithPatterns = samples.map((sample) => {
    const pattern = extractPattern(get(sample, field) as string);
    return {
      document: sample,
      fullPattern: pattern,
      truncatedPattern: pattern.slice(0, 10),
      fieldValue: get(sample, field) as string,
    };
  });

  const patternsToProcess = orderBy(
    Object.values(
      mapValues(
        // group by truncated pattern
        groupBy(samplesWithPatterns, 'truncatedPattern'),
        (samplesForTruncatedPattern, truncatedPattern) => {
          // return truncated pattern, count, and unique example values, shuffled
          return {
            truncatedPattern,
            count: samplesForTruncatedPattern.length,
            exampleValues: shuffle(
              uniq(samplesForTruncatedPattern.map(({ fieldValue }) => fieldValue))
            ).slice(0, NUMBER_SAMPLES_PER_PATTERN),
          };
        }
      )
    ),
    'count',
    'desc'
  );
```

Not really concerned about LOC, but wondering if this improves readability

flash1293 replied:
Happy to go all in on lodash here :) I will rewrite and add some tests, thanks

dgieselaar commented on x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/suggestions_handler.ts:
```diff
@@ -0,0 +1,235 @@
+
+type SimulationWithPattern = ReturnType<typeof prepareSimulationResponse> & { pattern: string };
+
+async function processPattern(
```
do we need separate requests for this?

flash1293 replied:
Nastia did some experiments and I think this approach worked best. Happy to go with a single LLM request for now though if you think it would be better.

dgieselaar replied:
No that is fine, it's just that there is some cost and perf overhead per request, but lets stick with this for now

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274584)
* Commit: 4a27b4efeb54b44187d416ea268a9a6c74f5ba0a
* Kibana Serverless Image: `docker.elastic.co/kibana-ci/kibana-serverless:pr-208777-4a27b4efeb54`

### History
* :green_heart: [Build #274368](https://buildkite.com/elastic/kibana-pull-request/builds/274368) succeeded 0695680f05f93c0d217c5672077c89c21e0153d2
* :broken_heart: [Build #274356](https://buildkite.com/elastic/kibana-pull-request/builds/274356) failed 47dd7ed5d4dedcd58147dfd030ab78af8f4dad1e
* :broken_heart: [Build #274303](https://buildkite.com/elastic/kibana-pull-request/builds/274303) failed a81de65acb3659708cdab24bf354a6ca65c91516
* :broken_heart: [Build #274220](https://buildkite.com/elastic/kibana-pull-request/builds/274220) failed fa4b7608d1c2f8ac39b4f02ac6c375c9a1c25f6b
* :broken_heart: [Build #273545](https://buildkite.com/elastic/kibana-pull-request/builds/273545) failed c70cff03200ca3fc3a50f13d4c31b3f60e509a64

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274584","number":274584,"commit":"4a27b4efeb54b44187d416ea268a9a6c74f5ba0a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274368","number":274368,"commit":"0695680f05f93c0d217c5672077c89c21e0153d2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274356","number":274356,"commit":"47dd7ed5d4dedcd58147dfd030ab78af8f4dad1e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274303","number":274303,"commit":"a81de65acb3659708cdab24bf354a6ca65c91516"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274220","number":274220,"commit":"fa4b7608d1c2f8ac39b4f02ac6c375c9a1c25f6b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/273545","number":273545,"commit":"c70cff03200ca3fc3a50f13d4c31b3f60e509a64"}],"number":274584}
buildkite-pr-comment-->

flash1293 commented on x-pack/solutions/observability/plugins/streams_app/public/components/stream_detail_enrichment/hooks/use_processing_simulator.ts:
```diff
@@ -189,15 +220,41 @@ export const useProcessingSimulator = ({
+  const [selectedDocsFilter, setSelectedDocsFilter] =
+    useState<DocsFilterOption>('outcome_filter_all');
+
+  const filteredSamples = useMemo(() => {
```
This is actually quite important because it allows the user to limit the pattern generation to only the unmatched samples. This means they can generate patterns, accept some, fine tune them, then select the "unmatched" and only focus on them. A better state management absent, lifting it up is the only thing we can do.

Wdyt about exposing the filtered and unfiltered samples separately to make this clearer?

tonyghiani replied:
> Wdyt about exposing the filtered and unfiltered samples separately to make this clearer?

As a middle step makes sense at least to avoid confusion when getting back to this code

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274810)
* Commit: 80563af0803c92ad1cd6f8176433c3ce2cfc67a7

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/274810#0194f09d-5887-4fc6-a560-614e24ee6094)

### History
* :broken_heart: [Build #274748](https://buildkite.com/elastic/kibana-pull-request/builds/274748) failed f7230e112981f216c2263605b4ea28ba7c03e2c4
* :green_heart: [Build #274584](https://buildkite.com/elastic/kibana-pull-request/builds/274584) succeeded 4a27b4efeb54b44187d416ea268a9a6c74f5ba0a
* :green_heart: [Build #274368](https://buildkite.com/elastic/kibana-pull-request/builds/274368) succeeded 0695680f05f93c0d217c5672077c89c21e0153d2
* :broken_heart: [Build #274356](https://buildkite.com/elastic/kibana-pull-request/builds/274356) failed 47dd7ed5d4dedcd58147dfd030ab78af8f4dad1e
* :broken_heart: [Build #274303](https://buildkite.com/elastic/kibana-pull-request/builds/274303) failed a81de65acb3659708cdab24bf354a6ca65c91516
* :broken_heart: [Build #274220](https://buildkite.com/elastic/kibana-pull-request/builds/274220) failed fa4b7608d1c2f8ac39b4f02ac6c375c9a1c25f6b

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274810","number":274810,"commit":"80563af0803c92ad1cd6f8176433c3ce2cfc67a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274748","number":274748,"commit":"f7230e112981f216c2263605b4ea28ba7c03e2c4"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274584","number":274584,"commit":"4a27b4efeb54b44187d416ea268a9a6c74f5ba0a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274368","number":274368,"commit":"0695680f05f93c0d217c5672077c89c21e0153d2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274356","number":274356,"commit":"47dd7ed5d4dedcd58147dfd030ab78af8f4dad1e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274303","number":274303,"commit":"a81de65acb3659708cdab24bf354a6ca65c91516"}],"number":274810}
buildkite-pr-comment-->

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274835)
* Commit: 62f33cb164dccf5450ece8bf0dcae9e2a5ca3c91

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/274835#0194f0be-a7bc-48e3-b06d-02e65d1b95d2)

### History
* :broken_heart: [Build #274810](https://buildkite.com/elastic/kibana-pull-request/builds/274810) failed 80563af0803c92ad1cd6f8176433c3ce2cfc67a7
* :broken_heart: [Build #274748](https://buildkite.com/elastic/kibana-pull-request/builds/274748) failed f7230e112981f216c2263605b4ea28ba7c03e2c4
* :green_heart: [Build #274584](https://buildkite.com/elastic/kibana-pull-request/builds/274584) succeeded 4a27b4efeb54b44187d416ea268a9a6c74f5ba0a
* :green_heart: [Build #274368](https://buildkite.com/elastic/kibana-pull-request/builds/274368) succeeded 0695680f05f93c0d217c5672077c89c21e0153d2
* :broken_heart: [Build #274356](https://buildkite.com/elastic/kibana-pull-request/builds/274356) failed 47dd7ed5d4dedcd58147dfd030ab78af8f4dad1e

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274835","number":274835,"commit":"62f33cb164dccf5450ece8bf0dcae9e2a5ca3c91"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274810","number":274810,"commit":"80563af0803c92ad1cd6f8176433c3ce2cfc67a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274748","number":274748,"commit":"f7230e112981f216c2263605b4ea28ba7c03e2c4"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274584","number":274584,"commit":"4a27b4efeb54b44187d416ea268a9a6c74f5ba0a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274368","number":274368,"commit":"0695680f05f93c0d217c5672077c89c21e0153d2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274356","number":274356,"commit":"47dd7ed5d4dedcd58147dfd030ab78af8f4dad1e"}],"number":274835}
buildkite-pr-comment-->

tonyghiani commented on x-pack/solutions/observability/plugins/streams_app/public/components/stream_detail_enrichment/hooks/use_processing_simulator.ts:
```diff
@@ -146,6 +147,14 @@ export const useProcessingSimulator = ({
 
+  const [memoizedSamplingCondition, setMemoizedSamplingCondition] = useState(samplingCondition);
+
+  useDeepCompareEffect(() => {
+    if (samplingCondition !== memoizedSamplingCondition) {
+      setMemoizedSamplingCondition(samplingCondition);
+    }
+  }, [samplingCondition]);
```
I don't think we need a deep compare effect, you can run a similar comparison at render time and skip this extra side-effect:
```tsx
const [memoizedSamplingCondition, setMemoizedSamplingCondition] = useState(samplingCondition);
if (!isEqual(samplingCondition, memoizedSamplingCondition) {
  setMemoizedSamplingCondition(samplingCondition);
}
```

dgieselaar replied:
maybe a stupid question, but why not calculate `samplingCondition` in the `useStreamsAppFetch` callback?

tonyghiani replied:
Because the `samplingCondition` is derived from the `liveProcessors`, Joe added this memoization probably because the condition was recomputed on each live processor change, even if the condition didn't change at all.
If we calculate that inside the `useStreamsAppFetch` call, it would cause the same behaviour Joe wants to fix.
Good catch Joe!

flash1293 replied:
Well, we want to refetch samples if the condition actually changes, but in the old state it would update every time the `liveProcessors` change, which includes the user typing a pattern, which is not what we want.

dgieselaar replied:
ah right, I see the change now, I thought the suggestion was to get rid of the deep equals comparison entirely.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/275066)
* Commit: 6659dfc7fdc9ffa2d5b4052babce7f22d6a830c4
* Kibana Serverless Image: `docker.elastic.co/kibana-ci/kibana-serverless:pr-208777-6659dfc7fdc9`

### Failed CI Steps
* [x-pack/test/alerting_api_integration/spaces_only/tests/alerting/group1/config.ts](https://buildkite.com/elastic/kibana-pull-request/builds/275066#0194f4d3-807f-4d23-9b90-70289303ee93)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/275066#0194f4d3-807f-4d23-9b90-70289303ee93) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/275066/jobs/0194f4d3-807f-4d23-9b90-70289303ee93/artifacts/0194f4e2-9781-47b3-95ae-ebfd24ed0762) x-pack/test/alerting_api_integration/spaces_only/tests/alerting/group1/config.ts / Alerting getActionErrorLog get and filter action error logs for rules with multiple action errors

### History
* :broken_heart: [Build #275042](https://buildkite.com/elastic/kibana-pull-request/builds/275042) failed 996573eb0932f6d7ac75b9fb789f32fad0483b45
* :broken_heart: [Build #274835](https://buildkite.com/elastic/kibana-pull-request/builds/274835) failed 62f33cb164dccf5450ece8bf0dcae9e2a5ca3c91
* :broken_heart: [Build #274810](https://buildkite.com/elastic/kibana-pull-request/builds/274810) failed 80563af0803c92ad1cd6f8176433c3ce2cfc67a7
* :broken_heart: [Build #274748](https://buildkite.com/elastic/kibana-pull-request/builds/274748) failed f7230e112981f216c2263605b4ea28ba7c03e2c4
* :green_heart: [Build #274584](https://buildkite.com/elastic/kibana-pull-request/builds/274584) succeeded 4a27b4efeb54b44187d416ea268a9a6c74f5ba0a

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/275066","number":275066,"commit":"6659dfc7fdc9ffa2d5b4052babce7f22d6a830c4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/275042","number":275042,"commit":"996573eb0932f6d7ac75b9fb789f32fad0483b45"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274835","number":274835,"commit":"62f33cb164dccf5450ece8bf0dcae9e2a5ca3c91"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274810","number":274810,"commit":"80563af0803c92ad1cd6f8176433c3ce2cfc67a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274748","number":274748,"commit":"f7230e112981f216c2263605b4ea28ba7c03e2c4"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274584","number":274584,"commit":"4a27b4efeb54b44187d416ea268a9a6c74f5ba0a"}],"number":275066}
buildkite-pr-comment-->

flash1293 commented:
@tonyghiani @dgieselaar I think this is ready for another look

tonyghiani commented on x-pack/solutions/observability/plugins/streams_app/public/components/stream_detail_enrichment/processors/grok/grok_ai_suggestions.tsx:
```diff
@@ -0,0 +1,334 @@
+    prefix: 'splitButtonPopover',
+  });
+
+  const onButtonClick = () => {
+    togglePopover(!isPopoverOpen);
+  };
```
Switching to useBoolean meant we don't need to re-declare this function, but you can directly pass the `togglePopover` handler to the onClick event. This can be removed

tonyghiani commented on x-pack/solutions/observability/plugins/streams_app/public/components/stream_detail_enrichment/processors/grok/index.tsx:
```diff
@@ -20,6 +20,7 @@ export const GrokProcessorForm = () => {
       <ProcessorFieldSelector />
       <GrokPatternsEditor />
       <EuiSpacer size="m" />
+      <EuiSpacer size="m" />
```
Probably you don't need a new spacer but just update the size of the one already present above

tonyghiani reviewed (commented):
I gave a second pass on the PR, overall looks good, I left some suggestions around.
Apart from that, I found 2 behaviours which might need updates:

**Pattern validation stuck**
When the pattern input is manually cleared and then a suggestion is added, the input validation is not retriggered, leaving an invalid state callout in the input.

https://github.com/user-attachments/assets/9e47fd00-e4d8-42f3-a729-5a0a21ad357b

**No feedback on unretrieved suggestions**
Is it expected to not give any feedback when no suggestions are detected? It might feel a bit broken from the user perspective, as they won't know if the pattern generation worked or no.

https://github.com/user-attachments/assets/2225dc63-01b5-46e1-8e37-2e28fc23719e

**No way to enable suggestions**
If the user doesn't explicitly have a connector configuration, they won't even now of this capability, which might be an attracting reason for them to enable it.
Should we handle this somehow? cc @LucaWintergerst 
P.S. The spacing in the grok editor seems a bit off when the suggestions are disabled.

<img width="415" alt="Screenshot 2025-02-12 at 10 23 09" src="https://github.com/user-attachments/assets/37753ea5-684d-48dd-a210-ee0f37394315" />

dgieselaar commented on x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/suggestions_handler.ts:
```diff
@@ -0,0 +1,211 @@
+      `,
+    schema: {
+      type: 'object',
+      properties: {
```
you can use `required` here if you have any properties that should be

dgieselaar commented on x-pack/solutions/observability/plugins/streams_app/public/components/stream_detail_enrichment/enrichment_context.tsx:
```diff
@@ -0,0 +1,46 @@
+  definition: IngestStreamGetResponse;
+}
+
+export const context = createContext<StreamEnrichmentContext | undefined>(undefined);
+
+export function StreamsEnrichmentContextProvider({
+  definition,
+  processingSimulator,
+
+  children,
+}: {
+  processingSimulator: UseProcessingSimulatorReturn;
+  definition: IngestStreamGetResponse;
+  children: React.ReactNode;
+}) {
+  const contextValue = useMemo(() => {
+    return {
+      definition,
+      ...processingSimulator,
+    };
+  }, [definition, processingSimulator]);
+  return <context.Provider value={contextValue}>{children}</context.Provider>;
+}
+
+export function useStreamsEnrichmentContext() {
+  const ctx = React.useContext(context);
+  if (!ctx) {
+    throw new Error(
+      'useStreamsEnrichmentContext must be used within a StreamsEnrichmentContextProvider'
+    );
+  }
+  return ctx;
+}
```
Looks like `constate` does not throw if the tree is not wrapped (https://github.com/diegohaz/constate/blob/bbad5665d124c64f71cadbd6ad44d95a96b02d4d/src/index.tsx#L38). Not a big fan of that, also because it's an empty object and not an undefined value (the latter of which will lead to runtime errors earlier instead of false negatives).

tonyghiani replied:
Didn't know it doesn't throw, then nevermind, the context check seems more solid

LucaWintergerst commented:
> If the user doesn't explicitly have a connector configuration, they won't even now of this capability, which might be an attracting reason for them to enable it.
Good call, let me work with Patri to get this added later 👍  


@patpscal - I am not sure if we have any prior implementations of this behaviour

We would have three states: 
License Basic: button hidden 
License Enterprise, no connector: button disabled (maybe?), link to connector setup  (but where? Do we use the place where the button usually is and link?)
License Enterprise, existing connector: button enabled

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/277327)
* Commit: 1740676a41da2912aa87ed4fd4f7ea6182aa8f53

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/277327#019519ca-c0ed-42bf-90ff-7f5ebdddc020)

### History
* :yellow_heart: [Build #275066](https://buildkite.com/elastic/kibana-pull-request/builds/275066) was flaky 6659dfc7fdc9ffa2d5b4052babce7f22d6a830c4
* :broken_heart: [Build #275042](https://buildkite.com/elastic/kibana-pull-request/builds/275042) failed 996573eb0932f6d7ac75b9fb789f32fad0483b45
* :broken_heart: [Build #274835](https://buildkite.com/elastic/kibana-pull-request/builds/274835) failed 62f33cb164dccf5450ece8bf0dcae9e2a5ca3c91
* :broken_heart: [Build #274810](https://buildkite.com/elastic/kibana-pull-request/builds/274810) failed 80563af0803c92ad1cd6f8176433c3ce2cfc67a7
* :broken_heart: [Build #274748](https://buildkite.com/elastic/kibana-pull-request/builds/274748) failed f7230e112981f216c2263605b4ea28ba7c03e2c4

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/277327","number":277327,"commit":"1740676a41da2912aa87ed4fd4f7ea6182aa8f53"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/275066","number":275066,"commit":"6659dfc7fdc9ffa2d5b4052babce7f22d6a830c4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/275042","number":275042,"commit":"996573eb0932f6d7ac75b9fb789f32fad0483b45"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274835","number":274835,"commit":"62f33cb164dccf5450ece8bf0dcae9e2a5ca3c91"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274810","number":274810,"commit":"80563af0803c92ad1cd6f8176433c3ce2cfc67a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274748","number":274748,"commit":"f7230e112981f216c2263605b4ea28ba7c03e2c4"}],"number":277327}
buildkite-pr-comment-->

afharo reviewed (approved):
`flattenObjectNestedLast` LGTM

flash1293 commented:
@tonyghiani I think I addressed all points raised except for the advertisement of the AI feature if it's not enabled. I'm going to add that in a follow-up PR


This is how the screen looks when no suggestions could be found:
<img width="402" alt="Screenshot 2025-02-18 at 17 40 18" src="https://github.com/user-attachments/assets/d155907d-ac6c-498f-a7f3-9111fad5db28" />

flash1293 commented:
You probably want to wait with another round of review until https://github.com/elastic/kibana/pull/209985 is merged (I pulled it in here and the diffs are mixed now)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/277364)
* Commit: 288dc3fa5edcb565830357f9ba3df2fc134db975

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/277364#01951a06-4b8a-4fe6-8a34-da150d6918de)

### History
* :broken_heart: [Build #277327](https://buildkite.com/elastic/kibana-pull-request/builds/277327) failed 1740676a41da2912aa87ed4fd4f7ea6182aa8f53
* :yellow_heart: [Build #275066](https://buildkite.com/elastic/kibana-pull-request/builds/275066) was flaky 6659dfc7fdc9ffa2d5b4052babce7f22d6a830c4
* :broken_heart: [Build #275042](https://buildkite.com/elastic/kibana-pull-request/builds/275042) failed 996573eb0932f6d7ac75b9fb789f32fad0483b45
* :broken_heart: [Build #274835](https://buildkite.com/elastic/kibana-pull-request/builds/274835) failed 62f33cb164dccf5450ece8bf0dcae9e2a5ca3c91
* :broken_heart: [Build #274810](https://buildkite.com/elastic/kibana-pull-request/builds/274810) failed 80563af0803c92ad1cd6f8176433c3ce2cfc67a7

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/277364","number":277364,"commit":"288dc3fa5edcb565830357f9ba3df2fc134db975"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/277327","number":277327,"commit":"1740676a41da2912aa87ed4fd4f7ea6182aa8f53"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/275066","number":275066,"commit":"6659dfc7fdc9ffa2d5b4052babce7f22d6a830c4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/275042","number":275042,"commit":"996573eb0932f6d7ac75b9fb789f32fad0483b45"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274835","number":274835,"commit":"62f33cb164dccf5450ece8bf0dcae9e2a5ca3c91"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274810","number":274810,"commit":"80563af0803c92ad1cd6f8176433c3ce2cfc67a7"}],"number":277364}
buildkite-pr-comment-->

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/277730)
* Commit: 44cba3c49be7a30c559ababd86a6e942e3e5df63

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/277730#01951ebf-bf1f-4e66-b3d4-3c8074ca6ca5)

### History
* :broken_heart: [Build #277364](https://buildkite.com/elastic/kibana-pull-request/builds/277364) failed 288dc3fa5edcb565830357f9ba3df2fc134db975
* :broken_heart: [Build #277327](https://buildkite.com/elastic/kibana-pull-request/builds/277327) failed 1740676a41da2912aa87ed4fd4f7ea6182aa8f53
* :yellow_heart: [Build #275066](https://buildkite.com/elastic/kibana-pull-request/builds/275066) was flaky 6659dfc7fdc9ffa2d5b4052babce7f22d6a830c4
* :broken_heart: [Build #275042](https://buildkite.com/elastic/kibana-pull-request/builds/275042) failed 996573eb0932f6d7ac75b9fb789f32fad0483b45
* :broken_heart: [Build #274835](https://buildkite.com/elastic/kibana-pull-request/builds/274835) failed 62f33cb164dccf5450ece8bf0dcae9e2a5ca3c91

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/277730","number":277730,"commit":"44cba3c49be7a30c559ababd86a6e942e3e5df63"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/277364","number":277364,"commit":"288dc3fa5edcb565830357f9ba3df2fc134db975"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/277327","number":277327,"commit":"1740676a41da2912aa87ed4fd4f7ea6182aa8f53"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/275066","number":275066,"commit":"6659dfc7fdc9ffa2d5b4052babce7f22d6a830c4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/275042","number":275042,"commit":"996573eb0932f6d7ac75b9fb789f32fad0483b45"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274835","number":274835,"commit":"62f33cb164dccf5450ece8bf0dcae9e2a5ca3c91"}],"number":277730}
buildkite-pr-comment-->

flash1293 commented:
@tonyghiani should be rebased with main - there are two things missing that I will address in a follow-up:
* Unit tests for the suggestions handler
* CTA if no LLM connector is available, but the user has permissions to configure it

tonyghiani reviewed (approved):
Looks good, there are some client side parts that will probably change with the state management refactor, but I'll handle that once I rebase this work into my WIP changes.

Agree on having some API test for the suggestions, there is a lot of logic going on there and having a test safety guard seems very necessary.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/277763)
* Commit: 6b3ce93ddc784a09419cc58dca596dcc3bfcbc5c
* Kibana Serverless Image: `docker.elastic.co/kibana-ci/kibana-serverless:pr-208777-6b3ce93ddc78`

### History
* :broken_heart: [Build #277730](https://buildkite.com/elastic/kibana-pull-request/builds/277730) failed 44cba3c49be7a30c559ababd86a6e942e3e5df63
* :broken_heart: [Build #277364](https://buildkite.com/elastic/kibana-pull-request/builds/277364) failed 288dc3fa5edcb565830357f9ba3df2fc134db975
* :broken_heart: [Build #277327](https://buildkite.com/elastic/kibana-pull-request/builds/277327) failed 1740676a41da2912aa87ed4fd4f7ea6182aa8f53
* :yellow_heart: [Build #275066](https://buildkite.com/elastic/kibana-pull-request/builds/275066) was flaky 6659dfc7fdc9ffa2d5b4052babce7f22d6a830c4
* :broken_heart: [Build #275042](https://buildkite.com/elastic/kibana-pull-request/builds/275042) failed 996573eb0932f6d7ac75b9fb789f32fad0483b45
* :broken_heart: [Build #274835](https://buildkite.com/elastic/kibana-pull-request/builds/274835) failed 62f33cb164dccf5450ece8bf0dcae9e2a5ca3c91

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/277763","number":277763,"commit":"6b3ce93ddc784a09419cc58dca596dcc3bfcbc5c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/277730","number":277730,"commit":"44cba3c49be7a30c559ababd86a6e942e3e5df63"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/277364","number":277364,"commit":"288dc3fa5edcb565830357f9ba3df2fc134db975"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/277327","number":277327,"commit":"1740676a41da2912aa87ed4fd4f7ea6182aa8f53"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/275066","number":275066,"commit":"6659dfc7fdc9ffa2d5b4052babce7f22d6a830c4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/275042","number":275042,"commit":"996573eb0932f6d7ac75b9fb789f32fad0483b45"}],"number":277763}
buildkite-pr-comment-->

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/13430306734

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.x|**Backport failed because of merge conflicts**<br><br>You might need to backport the following PRs to 8.x:<br> - [[streams] lifecycle - ingestion and total docs metadata (#210301)](https://github.com/elastic/kibana/pull/210301)|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 208777
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
